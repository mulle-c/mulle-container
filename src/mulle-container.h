//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
