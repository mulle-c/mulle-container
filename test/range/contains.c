#include <mulle-container/mulle-container.h>

#include <stdio.h>


static uintptr_t   values[6] =
{
   0,
   1,
   2,
   mulle_range_max,
   mulle_not_found_e,
   -1
};


static void   print_value( uintptr_t value)
{
   if( value == mulle_range_max)
      printf( "mulle_range_max");
   else
      if( value == mulle_not_found_e)
         printf( "mulle_not_found_e");
      else
         printf( "%ld", (long) value);
}




int   main( int argc, char *argv[])
{
   struct mulle_range   range1;
   struct mulle_range   range2;
   unsigned int         i, j;
   unsigned int         k, l;

   for( l = 0; l < 6; l++)
   {
      for( k = 0; k < 6; k++)
      {
         for( j= 0; j < 6; j++)
         {
            for( i = 0; i < 6; i++)
            {
               range1.location = values[ i];
               range1.length   = values[ j];
               range2.location = values[ k];
               range2.length   = values[ l];


               if( ! mulle_range_is_valid( range1))
                  continue;
               if( ! mulle_range_is_valid( range2))
                  continue;

               printf( "[");
               print_value( range1.location);
               printf( " - ");
               print_value( range1.length);
               printf( "] contains [");
               print_value( range2.location);
               printf( " - ");
               print_value( range2.length);
               printf( "] : %s\n",
                  mulle_range_contains( range1, range2) ? "YES" : "NO");
            }
         }
         printf( "\n");
      }
      printf( "\n");
   }
   return( 0);
}
