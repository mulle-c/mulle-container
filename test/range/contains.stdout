[0 - 0] contains [0 - 0] : NO
[1 - 0] contains [0 - 0] : NO
[2 - 0] contains [0 - 0] : NO
[mulle_range_max - 0] contains [0 - 0] : NO
[0 - 1] contains [0 - 0] : YES
[1 - 1] contains [0 - 0] : NO
[2 - 1] contains [0 - 0] : NO
[mulle_range_max - 1] contains [0 - 0] : NO
[0 - 2] contains [0 - 0] : YES
[1 - 2] contains [0 - 0] : NO
[2 - 2] contains [0 - 0] : NO
[0 - mulle_range_max] contains [0 - 0] : YES
[1 - mulle_range_max] contains [0 - 0] : NO
[0 - mulle_not_found_e] contains [0 - 0] : YES

[0 - 0] contains [1 - 0] : NO
[1 - 0] contains [1 - 0] : NO
[2 - 0] contains [1 - 0] : NO
[mulle_range_max - 0] contains [1 - 0] : NO
[0 - 1] contains [1 - 0] : NO
[1 - 1] contains [1 - 0] : YES
[2 - 1] contains [1 - 0] : NO
[mulle_range_max - 1] contains [1 - 0] : NO
[0 - 2] contains [1 - 0] : YES
[1 - 2] contains [1 - 0] : YES
[2 - 2] contains [1 - 0] : NO
[0 - mulle_range_max] contains [1 - 0] : YES
[1 - mulle_range_max] contains [1 - 0] : YES
[0 - mulle_not_found_e] contains [1 - 0] : YES

[0 - 0] contains [2 - 0] : NO
[1 - 0] contains [2 - 0] : NO
[2 - 0] contains [2 - 0] : NO
[mulle_range_max - 0] contains [2 - 0] : NO
[0 - 1] contains [2 - 0] : NO
[1 - 1] contains [2 - 0] : NO
[2 - 1] contains [2 - 0] : YES
[mulle_range_max - 1] contains [2 - 0] : NO
[0 - 2] contains [2 - 0] : NO
[1 - 2] contains [2 - 0] : YES
[2 - 2] contains [2 - 0] : YES
[0 - mulle_range_max] contains [2 - 0] : YES
[1 - mulle_range_max] contains [2 - 0] : YES
[0 - mulle_not_found_e] contains [2 - 0] : YES

[0 - 0] contains [mulle_range_max - 0] : NO
[1 - 0] contains [mulle_range_max - 0] : NO
[2 - 0] contains [mulle_range_max - 0] : NO
[mulle_range_max - 0] contains [mulle_range_max - 0] : NO
[0 - 1] contains [mulle_range_max - 0] : NO
[1 - 1] contains [mulle_range_max - 0] : NO
[2 - 1] contains [mulle_range_max - 0] : NO
[mulle_range_max - 1] contains [mulle_range_max - 0] : YES
[0 - 2] contains [mulle_range_max - 0] : NO
[1 - 2] contains [mulle_range_max - 0] : NO
[2 - 2] contains [mulle_range_max - 0] : NO
[0 - mulle_range_max] contains [mulle_range_max - 0] : NO
[1 - mulle_range_max] contains [mulle_range_max - 0] : YES
[0 - mulle_not_found_e] contains [mulle_range_max - 0] : YES




[0 - 0] contains [0 - 1] : NO
[1 - 0] contains [0 - 1] : NO
[2 - 0] contains [0 - 1] : NO
[mulle_range_max - 0] contains [0 - 1] : NO
[0 - 1] contains [0 - 1] : YES
[1 - 1] contains [0 - 1] : NO
[2 - 1] contains [0 - 1] : NO
[mulle_range_max - 1] contains [0 - 1] : NO
[0 - 2] contains [0 - 1] : YES
[1 - 2] contains [0 - 1] : NO
[2 - 2] contains [0 - 1] : NO
[0 - mulle_range_max] contains [0 - 1] : YES
[1 - mulle_range_max] contains [0 - 1] : NO
[0 - mulle_not_found_e] contains [0 - 1] : YES

[0 - 0] contains [1 - 1] : NO
[1 - 0] contains [1 - 1] : NO
[2 - 0] contains [1 - 1] : NO
[mulle_range_max - 0] contains [1 - 1] : NO
[0 - 1] contains [1 - 1] : NO
[1 - 1] contains [1 - 1] : YES
[2 - 1] contains [1 - 1] : NO
[mulle_range_max - 1] contains [1 - 1] : NO
[0 - 2] contains [1 - 1] : YES
[1 - 2] contains [1 - 1] : YES
[2 - 2] contains [1 - 1] : NO
[0 - mulle_range_max] contains [1 - 1] : YES
[1 - mulle_range_max] contains [1 - 1] : YES
[0 - mulle_not_found_e] contains [1 - 1] : YES

[0 - 0] contains [2 - 1] : NO
[1 - 0] contains [2 - 1] : NO
[2 - 0] contains [2 - 1] : NO
[mulle_range_max - 0] contains [2 - 1] : NO
[0 - 1] contains [2 - 1] : NO
[1 - 1] contains [2 - 1] : NO
[2 - 1] contains [2 - 1] : YES
[mulle_range_max - 1] contains [2 - 1] : NO
[0 - 2] contains [2 - 1] : NO
[1 - 2] contains [2 - 1] : YES
[2 - 2] contains [2 - 1] : YES
[0 - mulle_range_max] contains [2 - 1] : YES
[1 - mulle_range_max] contains [2 - 1] : YES
[0 - mulle_not_found_e] contains [2 - 1] : YES

[0 - 0] contains [mulle_range_max - 1] : NO
[1 - 0] contains [mulle_range_max - 1] : NO
[2 - 0] contains [mulle_range_max - 1] : NO
[mulle_range_max - 0] contains [mulle_range_max - 1] : NO
[0 - 1] contains [mulle_range_max - 1] : NO
[1 - 1] contains [mulle_range_max - 1] : NO
[2 - 1] contains [mulle_range_max - 1] : NO
[mulle_range_max - 1] contains [mulle_range_max - 1] : YES
[0 - 2] contains [mulle_range_max - 1] : NO
[1 - 2] contains [mulle_range_max - 1] : NO
[2 - 2] contains [mulle_range_max - 1] : NO
[0 - mulle_range_max] contains [mulle_range_max - 1] : NO
[1 - mulle_range_max] contains [mulle_range_max - 1] : YES
[0 - mulle_not_found_e] contains [mulle_range_max - 1] : YES




[0 - 0] contains [0 - 2] : NO
[1 - 0] contains [0 - 2] : NO
[2 - 0] contains [0 - 2] : NO
[mulle_range_max - 0] contains [0 - 2] : NO
[0 - 1] contains [0 - 2] : NO
[1 - 1] contains [0 - 2] : NO
[2 - 1] contains [0 - 2] : NO
[mulle_range_max - 1] contains [0 - 2] : NO
[0 - 2] contains [0 - 2] : YES
[1 - 2] contains [0 - 2] : NO
[2 - 2] contains [0 - 2] : NO
[0 - mulle_range_max] contains [0 - 2] : YES
[1 - mulle_range_max] contains [0 - 2] : NO
[0 - mulle_not_found_e] contains [0 - 2] : YES

[0 - 0] contains [1 - 2] : NO
[1 - 0] contains [1 - 2] : NO
[2 - 0] contains [1 - 2] : NO
[mulle_range_max - 0] contains [1 - 2] : NO
[0 - 1] contains [1 - 2] : NO
[1 - 1] contains [1 - 2] : NO
[2 - 1] contains [1 - 2] : NO
[mulle_range_max - 1] contains [1 - 2] : NO
[0 - 2] contains [1 - 2] : NO
[1 - 2] contains [1 - 2] : YES
[2 - 2] contains [1 - 2] : NO
[0 - mulle_range_max] contains [1 - 2] : YES
[1 - mulle_range_max] contains [1 - 2] : YES
[0 - mulle_not_found_e] contains [1 - 2] : YES

[0 - 0] contains [2 - 2] : NO
[1 - 0] contains [2 - 2] : NO
[2 - 0] contains [2 - 2] : NO
[mulle_range_max - 0] contains [2 - 2] : NO
[0 - 1] contains [2 - 2] : NO
[1 - 1] contains [2 - 2] : NO
[2 - 1] contains [2 - 2] : NO
[mulle_range_max - 1] contains [2 - 2] : NO
[0 - 2] contains [2 - 2] : NO
[1 - 2] contains [2 - 2] : NO
[2 - 2] contains [2 - 2] : YES
[0 - mulle_range_max] contains [2 - 2] : YES
[1 - mulle_range_max] contains [2 - 2] : YES
[0 - mulle_not_found_e] contains [2 - 2] : YES





[0 - 0] contains [0 - mulle_range_max] : NO
[1 - 0] contains [0 - mulle_range_max] : NO
[2 - 0] contains [0 - mulle_range_max] : NO
[mulle_range_max - 0] contains [0 - mulle_range_max] : NO
[0 - 1] contains [0 - mulle_range_max] : NO
[1 - 1] contains [0 - mulle_range_max] : NO
[2 - 1] contains [0 - mulle_range_max] : NO
[mulle_range_max - 1] contains [0 - mulle_range_max] : NO
[0 - 2] contains [0 - mulle_range_max] : NO
[1 - 2] contains [0 - mulle_range_max] : NO
[2 - 2] contains [0 - mulle_range_max] : NO
[0 - mulle_range_max] contains [0 - mulle_range_max] : YES
[1 - mulle_range_max] contains [0 - mulle_range_max] : NO
[0 - mulle_not_found_e] contains [0 - mulle_range_max] : YES

[0 - 0] contains [1 - mulle_range_max] : NO
[1 - 0] contains [1 - mulle_range_max] : NO
[2 - 0] contains [1 - mulle_range_max] : NO
[mulle_range_max - 0] contains [1 - mulle_range_max] : NO
[0 - 1] contains [1 - mulle_range_max] : NO
[1 - 1] contains [1 - mulle_range_max] : NO
[2 - 1] contains [1 - mulle_range_max] : NO
[mulle_range_max - 1] contains [1 - mulle_range_max] : NO
[0 - 2] contains [1 - mulle_range_max] : NO
[1 - 2] contains [1 - mulle_range_max] : NO
[2 - 2] contains [1 - mulle_range_max] : NO
[0 - mulle_range_max] contains [1 - mulle_range_max] : NO
[1 - mulle_range_max] contains [1 - mulle_range_max] : YES
[0 - mulle_not_found_e] contains [1 - mulle_range_max] : YES






[0 - 0] contains [0 - mulle_not_found_e] : NO
[1 - 0] contains [0 - mulle_not_found_e] : NO
[2 - 0] contains [0 - mulle_not_found_e] : NO
[mulle_range_max - 0] contains [0 - mulle_not_found_e] : NO
[0 - 1] contains [0 - mulle_not_found_e] : NO
[1 - 1] contains [0 - mulle_not_found_e] : NO
[2 - 1] contains [0 - mulle_not_found_e] : NO
[mulle_range_max - 1] contains [0 - mulle_not_found_e] : NO
[0 - 2] contains [0 - mulle_not_found_e] : NO
[1 - 2] contains [0 - mulle_not_found_e] : NO
[2 - 2] contains [0 - mulle_not_found_e] : NO
[0 - mulle_range_max] contains [0 - mulle_not_found_e] : NO
[1 - mulle_range_max] contains [0 - mulle_not_found_e] : NO
[0 - mulle_not_found_e] contains [0 - mulle_not_found_e] : YES














